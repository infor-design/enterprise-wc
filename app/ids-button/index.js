import IdsButton from '../../src/ids-button/ids-button';
