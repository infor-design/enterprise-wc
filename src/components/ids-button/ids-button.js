import {
  IdsElement,
  customElement,
  mix,
  scss,
  attributes
} from '../../core';

import { IdsStringUtils } from '../../utils';

import {
  IdsEventsMixin,
  IdsColorVariantMixin,
  IdsLocaleMixin,
  IdsThemeMixin,
  IdsTooltipMixin
} from '../../mixins';

import { renderLoop, IdsRenderLoopItem } from '../ids-render-loop';
import '../ids-text';
import '../ids-icon';
import styles from './ids-button.scss';

const { stringToBool } = IdsStringUtils;

// Button Styles
const BUTTON_TYPES = [
  'default',
  'primary',
  'secondary',
  'tertiary',
  'destructive',
  'swipe-action-left',
  'swipe-action-right'
];

// Default Button state values
const BUTTON_DEFAULTS = {
  cssClass: [],
  disabled: false,
  tabIndex: 0,
  type: BUTTON_TYPES[0]
};

// Definable attributes
const BUTTON_ATTRIBUTES = [
  attributes.CSS_CLASS,
  attributes.DISABLED,
  attributes.ICON,
  attributes.ICON_ALIGN,
  attributes.ID,
  attributes.NO_PADDING,
  attributes.NO_RIPPLE,
  attributes.SQUARE,
  attributes.TEXT,
  attributes.TYPE,
  attributes.TABINDEX,
  attributes.COLOR_VARIANT
];

// Icon alignments
const ICON_ALIGN = [
  'align-icon-start',
  'align-icon-end'
];

const baseProtoClasses = [
  'ids-button',
  'ids-icon-button',
  'ids-menu-button',
  'ids-toggle-button'
];

/**
 * IDS Button Component
 * @type {IdsButton}
 * @inherits IdsElement
 * @mixes IdsThemeMixin
 * @mixes IdsEventsMixin
 * @mixes IdsTooltipMixin
 * @part button - the button element
 * @part icon - the icon element
 * @part text - the text element
 */
@customElement('ids-button')
@scss(styles)
class IdsButton extends mix(IdsElement).with(
    IdsEventsMixin,
    IdsColorVariantMixin,
    IdsLocaleMixin,
    IdsThemeMixin,
    IdsTooltipMixin
  ) {
  constructor() {
    super();
    this.state = {};
    Object.keys(BUTTON_DEFAULTS).forEach((prop) => {
      this.state[prop] = BUTTON_DEFAULTS[prop];
    });
    this.shouldUpdate = true;
  }

  /**
   * Override `attributeChangedCallback` from IdsElement to wrap its normal operation in a
   * check for a true `shouldUpdate` property.
   * @param  {string} name The property name
   * @param  {string} oldValue The property old value
   * @param  {string} newValue The property new value
   */
  attributeChangedCallback(name, oldValue, newValue) {
    if (this.shouldUpdate) {
      switch (name) {
      // Convert "tabindex" to "tabIndex"
      case 'tabindex':
        if (Number.isNaN(Number.parseInt(newValue))) {
          this.tabIndex = null;
        }
        this.tabIndex = Number(newValue);
        break;
      default:
        super.attributeChangedCallback.apply(this, [name, oldValue, newValue]);
        break;
      }
    }
  }

  /**
   * Button-level `connectedCallback` implementation
   * @returns {void}
   */
  connectedCallback() {
    this.attachEventHandlers();
    this.setIconAlignment();
    this.shouldUpdate = true;
    super.connectedCallback();
  }

  /**
   * Return the attributes we handle as getters/setters
   * @returns {Array} The attributes in an array
   */
  static get attributes() {
    return [...super.attributes, ...BUTTON_ATTRIBUTES];
  }

  /**
   * Inherited from `IdsColorVariantMixin`
   * @returns {Array<string>} List of available color variants for this component
   */
  colorVariants = ['alternate'];

  /**
   * Figure out the classes
   * @private
   * @readonly
   * @returns {Array} containing classes used to identify this button prototype
   */
  get protoClasses() {
    const textSlot = this.querySelector('span:not(.audible), ids-text:not([audible])');
    const iconSlot = this.querySelector('ids-icon[slot]') || this.querySelector('ids-icon');
    if (iconSlot && (!textSlot)) {
      return ['ids-icon-button'];
    }
    return ['ids-button'];
  }

  /**
   * Refreshes this button's prototype CSS class
   * @private
   * @returns {void}
   */
  refreshProtoClasses() {
    const cl = this.button.classList;
    const newProtoClass = this.protoClasses;

    cl.remove(...baseProtoClasses);
    cl.add(...newProtoClass);
  }

  /**
   * Inner template contents
   * @returns {string} The template
   */
  template() {
    let cssClass = '';
    let protoClasses = '';
    let disabled = '';
    let icon = '';
    let tabIndex = 'tabindex="0"';
    let text = '';
    let type = '';
    if (this.state?.cssClass) {
      cssClass = ` ${this.state.cssClass.join(' ')}`;
    }
    if (this.state?.disabled) {
      disabled = ` disabled="true"`;
    }
    if (this.state?.tabIndex) {
      tabIndex = `tabindex="${this.state.tabIndex}"`;
    }
    if (this.state?.icon) {
      icon = `<ids-icon slot="icon" part="icon" icon="${this.state.icon}"></ids-icon>`;
    }
    if (this.state?.text) {
      text = `<span slot="text" part="text">${this.state.text}</span>`;
    }
    if (this.state && this.state?.type !== 'default') {
      type = ` btn-${this.state.type}`;
    }

    if (this.hasAttribute(attributes.SQUARE)) {
      cssClass += ' square';
    }

    if (this.protoClasses.length) {
      protoClasses = `${this.protoClasses.join(' ')}`;
    }

    let alignCSS = ' align-icon-start';
    let namedSlots = `<slot name="icon" part="icon">${icon}</slot><slot name="text">${text}</slot>`;
    if (this.state?.iconAlign === 'end') {
      alignCSS = ' align-icon-end';
      namedSlots = `<slot name="text" part="text">${text}</slot><slot name="icon">${icon}</slot>`;
    }

    return `<button part="button" class="${protoClasses}${type}${alignCSS}${cssClass}" ${tabIndex}${disabled}>
      ${namedSlots}
      <slot>${icon}${text}</slot>
    </button>`;
  }

  /**
   * Sets up event listeners
   * @private
   * @returns {void}
   */
  attachEventHandlers() {
    let x;
    let y;
    let preceededByTouchstart = false;
    if (this.noRipple) {
      return;
    }

    this.onEvent('click.ripple', this.button, (e) => {
      if (preceededByTouchstart) {
        preceededByTouchstart = false;
        return;
      }
      x = e.clientX !== 0 ? e.clientX : undefined;
      y = e.clientY !== 0 ? e.clientY : undefined;
      this.createRipple(x, y);
    });

    this.onEvent('touchstart.ripple', this.button, (e) => {
      if (e.touches && e.touches.length > 0) {
        const touch = e.touches[0];
        x = touch.clientX !== 0 ? touch.clientX : undefined;
        y = touch.clientY !== 0 ? touch.clientY : undefined;
        this.createRipple(x, y);
        preceededByTouchstart = true;
      }
    }, {
      passive: true
    });

    // Respond to parent changing language
    this.offEvent('languagechange.button');
    this.onEvent('languagechange.button', this.closest('ids-container'), async (e) => {
      await this.setLanguage(e.detail.language.name);
      this.container.classList[this.locale.isRTL() ? 'add' : 'remove']('rtl');
    });
  }

  /**
   * @readonly
   * @returns {HTMLButtonElement} reference to the true button element used in the Shadow Root
   */
  get button() {
    return this.shadowRoot.querySelector('button');
  }

  /**
   * @param {Array|string} val containing CSS classes that will be applied to the button
   * Strings will be split into an array and separated by whitespace.
   */
  set cssClass(val) {
    let attr = val;
    let newCl = [];
    // @TODO replace with clone utils method
    const prevClasses = [].concat(this.state.cssClass);

    if (Array.isArray(val)) {
      newCl = val;
      attr = val.join(' ');
    } else if (typeof val === 'string' && val.length) {
      newCl = val.split(' ');
    }

    this.state.cssClass = newCl;
    if (newCl.length) {
      this.setAttribute(attributes.CSS_CLASS, attr.toString());
    } else {
      this.removeAttribute(attributes.CSS_CLASS);
    }

    // Remove/Set CSS classes on the actual inner Button component
    const buttonCl = this.button.classList;
    const buttonClArr = Array.from(buttonCl);
    prevClasses.forEach((cssClass) => {
      if (!newCl.includes(cssClass)) {
        buttonCl.remove(cssClass);
      }
    });
    newCl.forEach((newCssClass) => {
      if (!buttonClArr.includes(newCssClass)) {
        buttonCl.add(newCssClass);
      }
    });
  }

  get cssClass() {
    return this.state.cssClass;
  }

  /**
   * Passes a disabled attribute from the custom element to the button
   * @param {boolean|string} val true if the button will be disabled
   */
  set disabled(val) {
    const isValueTruthy = stringToBool(val);
    this.shouldUpdate = false;
    if (isValueTruthy) {
      this.setAttribute(attributes.DISABLED, '');
    } else {
      this.removeAttribute(attributes.DISABLED);
    }

    this.shouldUpdate = true;
    this.state.disabled = isValueTruthy;

    if (this.button) {
      this.button.disabled = isValueTruthy;
    }
  }

  get disabled() {
    return this.state.disabled;
  }

  /**
   * Passes a tabIndex attribute from the custom element to the button
   * @param {number} val the tabIndex value
   * @returns {void}
   */
  set tabIndex(val) {
    const trueVal = Number(val);

    if (Number.isNaN(trueVal) || trueVal < -1) {
      this.state.tabIndex = 0;
      this.button.setAttribute(attributes.TABINDEX, '0');
      this.removeAttribute(attributes.TABINDEX);
      return;
    }

    this.state.tabIndex = trueVal;
    this.button.setAttribute(attributes.TABINDEX, `${trueVal}`);
  }

  /**
   * @returns {number} the current tabIndex number for the button
   */
  get tabIndex() {
    return this.state.tabIndex;
  }

  /**
   * Sets the icon on the button
   * @param {undefined|string} val representing the icon to set
   */
  set icon(val) {
    if (typeof val !== 'string' || !val.length) {
      this.removeAttribute(attributes.ICON);
      this.state.icon = undefined;
      this.removeIcon();
      return;
    }
    this.state.icon = val;
    this.setAttribute(attributes.ICON, val);
    this.appendIcon(val);
  }

  /**
   * Gets the current icon used on the button
   * @returns {undefined|string} a defined IdsIcon's `icon` attribute, if one is present
   */
  get icon() {
    return this.querySelector('ids-icon')?.getAttribute('icon');
  }

  /**
   * Sets the alignment of an existing icon to the 'start' or 'end' of the text
   * @param {string} val the alignment type to set.
   */
  set iconAlign(val) {
    let trueVal = val;
    if (!ICON_ALIGN.includes(`align-icon-${val}`)) {
      trueVal = 'start';
    }
    this.state.iconAlign = trueVal;
    this.setIconAlignment();
  }

  /**
   * @returns {string} containing 'start' or 'end'
   */
  get iconAlign() {
    return this.state.iconAlign;
  }

  /**
   * Check if an icon exists, and adds the icon if it's missing
   * @param {string} iconName The icon name to check
   * @private
   */
  appendIcon(iconName) {
    // First look specifically for an icon slot.
    const icon = this.querySelector(`ids-icon`); // @TODO check for dropdown/expander icons here

    if (icon) {
      icon.icon = iconName;
      this.setIconAlignment();
    } else {
      this.insertAdjacentHTML('afterbegin', `<ids-icon slot="icon" icon="${iconName}" class="ids-icon"></ids-icon>`);
    }

    this.refreshProtoClasses();
  }

  /**
   * Check if an icon exists, and removes the icon if it's present
   * @private
   */
  removeIcon() {
    const icon = this.querySelector(`ids-icon`); // @TODO check for dropdown/expander icons here

    if (icon) {
      icon.remove();
    }
    this.setIconAlignment();
    this.refreshProtoClasses();
  }

  /**
   * Adds/Removes Icon Alignment CSS classes to/from the inner button component.
   * @private
   */
  setIconAlignment() {
    const alignment = this.iconAlign || 'start';
    const iconStr = this.icon;
    this.button.classList.remove(...ICON_ALIGN);

    // Append the icon, if needed
    if (iconStr) {
      this.button.classList.add(`align-icon-${alignment}`);
    }

    // Re-arrange the slots
    const iconSlot = this.button.querySelector('slot[name="icon"]');
    if (!iconSlot) {
      return;
    }

    if (alignment === 'end') {
      this.button.appendChild(iconSlot);
    } else {
      this.button.prepend(iconSlot);
    }
  }

  /**
   * @param {string} val the text value
   * @returns {void}
   */
  set text(val) {
    this.removeAttribute(attributes.TEXT);

    if (typeof val !== 'string' || !val.length) {
      this.state.text = '';
      this.removeText();
      return;
    }

    // @TODO: Run this through an XSS check
    this.state.text = val;
    this.appendText(val);
  }

  /**
   * @returns {string} the current text value
   */
  get text() {
    const textElem = this.querySelector('span:not(.audible)');
    if (textElem && textElem.textContent?.length) {
      return textElem.textContent;
    }
    return this.textContent;
  }

  /**
   * Check if the text slot exists, and appends it if it's missing
   * @param {string} val New text contents
   * @private
   */
  appendText(val) {
    const text = this.querySelector(`span:not(.audible)`);
    if (text) {
      text.textContent = val;
    } else {
      this.insertAdjacentHTML('afterbegin', `<span>${val}</span>`);
    }
    this.refreshProtoClasses();
  }

  /**
   * Checks if the text slot exists, and removes it if necessary
   * @private
   */
  removeText() {
    const text = this.querySelector(`span:not(.audible)`);
    if (text) {
      text.remove();
    }
    this.refreshProtoClasses();
  }

  /**
   * Set the button types between 'default', 'primary', 'secondary', 'tertiary', or 'destructive'
   * @param {string} val a valid button "type"
   */
  set type(val) {
    if (!val || BUTTON_TYPES.indexOf(val) <= 0) {
      this.removeAttribute(attributes.TYPE);
      this.state.type = BUTTON_TYPES[0];
    } else {
      this.setAttribute(attributes.TYPE, val);
      if (this.state.type !== val) {
        this.state.type = val;
      }
    }
    this.setTypeClass(val);
  }

  /**
   * @returns {string} the currently set type
   */
  get type() {
    return this.state.type;
  }

  /**
   * If set to true the ripple effect will be disabled.
   * @param {boolean} val The ripple value
   */
  set noRipple(val) {
    if (IdsStringUtils.stringToBool(val)) {
      this.setAttribute(attributes.NO_RIPPLE, true);
      this.state.noRipple = true;
      this.offEvent('click.ripple');
      this.offEvent('touchstart.ripple');
      return;
    }
    this.removeAttribute(attributes.NO_RIPPLE);
    this.state.noRipple = false;
  }

  /**
   * @returns {string} the currently set type
   */
  get noRipple() {
    return this.state.noRipple || false;
  }

  /**
   * @param {boolean} val true if the button should not have standard padding rules applied
   */
  set noPadding(val) {
    const isTruthy = this.noPadding;
    const trueVal = IdsStringUtils.stringToBool(val);
    if (isTruthy !== trueVal) {
      if (trueVal) {
        this.container.classList.add('no-padding');
        this.setAttribute('no-padding', '');
      } else {
        this.container.classList.remove('no-padding');
        this.removeAttribute('no-padding');
      }
    }
  }

  /**
   * @returns {boolean} true if the button does not currently have standard padding rules applied
   */
  get noPadding() {
    return this.container.classList.contains('no-padding');
  }

  /**
   * @param {boolean} value whether the corners of the button as an icon-button should be angled/90°
   */
  set square(value) {
    const isTruthy = stringToBool(value);

    if (isTruthy && !this.button.classList.contains('square')) {
      this.button.classList.add('square');
    } else if (!isTruthy && this.button.classList.contains('square')) {
      this.button.classList.remove('square');
    }

    if (isTruthy && !this.hasAttribute(attributes.SQUARE)) {
      this.setAttribute(attributes.SQUARE, '');
    } else if (!isTruthy && this.hasAttribute(attributes.SQUARE)) {
      this.removeAttribute(attributes.SQUARE);
    }
  }

  /**
   * @returns {boolean} whether the corners of the button as an icon-button are angled/90°
   */
  get square() {
    return this.hasAttribute(attributes.SQUARE);
  }

  /**
   * Sets the correct type class on the Shadow button.
   * @private
   * @param {string} val desired type class
   */
  setTypeClass(val) {
    BUTTON_TYPES.forEach((type) => {
      const typeClassName = `btn-${type}`;
      if (val === type) {
        if (type !== 'default' && !this.button.classList.contains(typeClassName)) {
          this.button.classList.add(typeClassName);
        }
        return;
      }
      if (this.button.classList.contains(typeClassName)) {
        this.button.classList.remove(typeClassName);
      }
    });
  }

  /**
   * The math used for getting the ripple offsets
   * @private
   * @param {number} x the X coordinate
   * @param {number} y the Y coordinate
   * @returns {object} containing x/y coordinates of the ripple
   */
  getRippleOffsets(x, y) {
    const btnRect = this.getBoundingClientRect();
    const halfRippleSize = this.button.classList.contains('ids-icon-button') ? 35 : 125;
    let btnX;
    let btnY;

    // If "X" is defined, assume it's page coordinates and subtract the
    // custom element's offsets from its location in the page.
    // Otherwise, simply set the offset to the center of the button.
    if (!x) {
      btnX = (btnRect.width / 2);
    } else {
      btnX = x - btnRect.left;
    }

    // If "Y" is defined, assume it's page coordinates and subtract the
    // custom element's offsets from its location in the page.
    // Otherwise, simply set the offset to the center of the button.
    if (!y) {
      btnY = (btnRect.height / 2);
    } else {
      btnY = y - btnRect.top;
    }

    // Subtract half the ripple size from each dimension.
    btnX -= halfRippleSize;
    btnY -= halfRippleSize;

    return { x: btnX, y: btnY };
  }

  /**
   * Generates a "ripple" effect on a specified location inside the button's boundaries.
   * The coordinates defined are actual page coordinates, using the top/left of the page as [0,0],
   * which allows this to connect easily to mouse/touch events.
   * @private
   * @param {number} x the X coordinate
   * @param {number} y the Y coordinate
   * @returns {void}
   */
  createRipple(x, y) {
    if (this.disabled) {
      return;
    }

    // Remove pre-existing ripples
    const otherRippleEls = this.button.querySelectorAll('.ripple-effect');
    otherRippleEls.forEach((rippleEl) => {
      rippleEl.remove();
    });

    // Make/Place a new ripple
    const rippleEl = document.createElement('span');
    const btnOffsets = this.getRippleOffsets(x, y);
    rippleEl.classList.add('ripple-effect');
    rippleEl.setAttribute('aria-hidden', 'true');
    rippleEl.setAttribute('focusable', 'false');

    this.button.prepend(rippleEl);
    rippleEl.style.left = `${btnOffsets.x}px`;
    rippleEl.style.top = `${btnOffsets.y}px`;
    rippleEl.classList.add('animating');

    // After a short time, remove the ripple effect
    if (this.rippleTimeout) {
      this.rippleTimeout.destroy(true);
    }

    this.rippleTimeout = renderLoop.register(new IdsRenderLoopItem({
      duration: 1200,
      timeoutCallback() {
        rippleEl.remove();
      }
    }));
  }

  /**
   * Overrides the standard "focus" behavior to instead pass focus to the inner HTMLButton element.
   */
  focus() {
    this.button.focus();
  }

  /**
   * Implements callback from IdsColorVariantMixin used to
   * update the color variant on children components
   * @returns {void}
   */
  onColorVariantRefresh() {
    const icons = this.querySelectorAll('ids-icon');
    const texts = this.querySelectorAll('ids-text');
    const iterator = (el) => {
      el.colorVariant = this.colorVariant;
    };
    [...icons, ...texts].forEach(iterator);
  }
}

export { IdsButton, BUTTON_ATTRIBUTES, BUTTON_TYPES };
