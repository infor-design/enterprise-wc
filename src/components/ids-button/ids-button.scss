/**
 * Ids Button Component CSS
 * (Includes standard and icon button rules)
 */
@import './ids-button-base';

:host {
  @include inline-flex();

  ::slotted(span) {
    vertical-align: bottom;
  }

  ::slotted(span),
  ::slotted(ids-text) {
    color: currentColor;
  }

  ::slotted(ids-icon) {
    vertical-align: middle;
    color: currentColor;
  }

  ::slotted(span.audible) {
    @include audible();
  }

  // Button elements should fill the space inside shadow root
  button {
    width: 100%;
  }
}

// Reset button display styles internally
button {
  display: block;
}

.ripple-effect {
  @include block();
  @include rounded-round();

  opacity: 0;
  position: absolute;
  transform: scale(0);

  &.animating {
    @include animation(ripple-animation 1.2s ease-out);
  }
}

// Applies to all non-icon button types
.ids-button,
.ids-toggle-button,
.ids-menu-button:not(.ids-icon-button),
.ids-modal-button {
  .ids-icon {
    @include inline-flex();

    place-self: center;
    vertical-align: middle;
  }

  .ripple-effect {
    height: 250px;
    width: 250px;
  }
}

.ids-button,
.ids-toggle-button,
.ids-menu-button {
  @include ids-base-button-styles();
  @include ids-standard-button-colors();

  // Technically not valid
  ::slotted(ids-icon:only-child) {
    @include block();
  }

  &:not(.no-padding):not(.ids-icon-button) {
    @include ids-standard-button-padding();
  }

  &.no-padding {
    @include ids-inline-button-padding();
  }
}

.ids-icon-button {
  @include ids-base-button-styles();
  @include ids-standard-button-colors();

  // Everything below is only for icon buttons
  @include px-8();
  @include py-8();

  &:not(.square) {
    @include rounded-round();
  }

  // Fixes an issue that causes the ripple effect not to appear "round" on Safari.
  // NOTE: Needs to be applied only when not focused because it cuts off the focus state.
  &:not(:focus) {
    mask-image: radial-gradient(var(--ids-color-palette-azure-10), var(--ids-color-palette-azure-10));
    /* stylelint-disable property-no-vendor-prefix */
    -webkit-mask-image: radial-gradient(var(--ids-color-palette-azure-10), var(--ids-color-palette-azure-10));
    /* stylelint-enable property-no-vendor-prefix */
  }

  .ids-icon {
    @include block();
    @include pl-0();

    font-size: 0;
  }

  span.audible {
    @include audible();
  }

  .ripple-effect {
    height: 70px;
    width: 70px;
  }
}

// Mix-in special rules for contextual buttons
.ids-button,
.ids-icon-button {
  @include ids-standard-contextual-button-colors();
}

// Specific Modifications for Modal Buttons
// Do not apply styles to modal close icon button, appearance as icon is required
.ids-modal-button:not(.ids-modal-icon-button) {
  @include ids-base-button-styles();
  @include ids-modal-button-colors();
  @include py-12();

  display: block;
  font-weight: 600;

  &:focus {
    z-index: 1;
  }
}

// Adjusting hitbox for the modal close icon
.ids-modal-icon-button {
  @include p-12();
}

@include keyframes(ripple-animation) {
  from {
    opacity: 0.3;
    transform: scale(0);
  }

  to {
    opacity: 0;
    transform: scale(2);
  }
}

// ====================================================
// Swipe Action Buttons (Swipe action component)
.btn-swipe-action-left,
.btn-swipe-action-right {
  @include bg-slate-30();
  @include text-slate-100();

  border-radius: 0;
  display: block;
  padding: 0 8px 8px;
  height: 100%;
  margin: 0;
  white-space: inherit;
  width: 85px;

  ::slotted(ids-icon) {
    display: block;
  }

  &:focus {
    box-shadow: none;
    border-color: transparent;
  }

  &:active {
    @include bg-slate-40();
  }

  &:hover {
    @include bg-slate-40();

    box-shadow: none;
    border-color: transparent;
  }
}

.btn-swipe-action-right {
  @include bg-azure-70();
  @include text-white();

  &:hover {
    @include bg-azure-80();
    @include text-white();
  }

  &:active {
    @include bg-azure-80();
  }
}

.btn-swipe-action-left[mode='dark'] {
  @include bg-slate-60();
}

.btn-swipe-action-right[mode='dark'] {
  @include bg-azure-50();
}

.btn-swipe-action-left[mode='contrast'] {
  @include bg-slate-30();
}

.btn-swipe-action-right[mode='contrast'] {
  @include bg-azure-50();
  @include text-white();
}

.btn-swipe-action-left[mode='light'][version='classic'] {
  @include bg-graphite-30();
}

.btn-swipe-action-right[mode='light'][version='classic'] {
  @include bg-azure-50();
  @include text-white();
}

.btn-swipe-action-left[mode='dark'][version='classic'] {
  @include bg-graphite-60();
}

.btn-swipe-action-right[mode='dark'][version='classic'] {
  @include bg-azure-60();
  @include text-white();
}

.btn-swipe-action-left[mode='contrast'][version='classic'] {
  @include bg-graphite-30();
}

.btn-swipe-action-right[mode='contrast'][version='classic'] {
  @include bg-azure-50();
  @include text-white();
}

// Non-RTL Styles
:not(.rtl) {
  // Applies to all non-icon button types
  &.ids-button,
  &.ids-toggle-button,
  &.ids-menu-button:not(.ids-icon-button),
  &.ids-modal-button {
    // Padding rules for buttons that insert both text/icon into the fallback (unnamed) slot.
    ::slotted(ids-icon:not(:first-child)), // shadowroot + text + icon
    ::slotted(span:nth-child(2):last-child), // shadowroot + icon + text
    ::slotted(ids-text:nth-child(2):last-child),
    ::slotted(span:nth-child(3):last-child), // shadowroot + icon + (audible text) + text
    ::slotted(ids-text:nth-child(3):last-child) {
      @include pl-4();
    }

    // Padding rules for standalone CSS buttons (match the rules/style above)
    .ids-icon:last-child,
    span + span,
    span:last-child {
      @include pl-4();
    }
  }
}

// RTL Styles
.rtl {
  // Applies to all non-icon button types
  &.ids-button,
  &.ids-toggle-button,
  &.ids-menu-button:not(.ids-icon-button),
  &.ids-modal-button {
    // Padding rules for buttons that insert both text/icon into the fallback (unnamed) slot.
    ::slotted(ids-icon:not(:first-child)), // shadowroot + text + icon
    ::slotted(span:nth-child(2):last-child), // shadowroot + icon + text
    ::slotted(ids-text:nth-child(2):last-child),
    ::slotted(span:nth-child(3):last-child), // shadowroot + icon + (audible text) + text
    ::slotted(ids-text:nth-child(3):last-child) {
      @include pr-4();

      padding-left: auto;
    }

    // Padding rules for standalone CSS buttons (match the rules/style above)
    .ids-icon:last-child,
    span + span,
    span:last-child {
      @include pr-4();

      padding-left: auto;
    }
  }
}
