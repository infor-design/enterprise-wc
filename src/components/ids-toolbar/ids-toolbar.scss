@import '../../core/ids-base';

:host {
  @include block();

  ::slotted(ids-toolbar-section[type='fluid']),
  ::slotted(ids-toolbar-section[type='title']),
  ::slotted(ids-toolbar-section[type='buttonset']) {
    flex-grow: 1;
  }

  &[disabled] {
    ::slotted(.ids-text) {
      color: text-slate-30();
    }
  }
}

.ids-toolbar {
  @include flex();

  align-items: center;
  flex-flow: row nowrap;
  height: 100%;
  justify-content: space-between;
  width: 100%;

  &.disabled {
    color: text-slate-30();

    .ids-text {
      color: inherit;
    }
  }
}
