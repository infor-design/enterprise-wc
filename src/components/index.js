import IdsAccordion from './ids-accordion';
import IdsAlert from './ids-alert';
import IdsAppMenu from './ids-app-menu';
import IdsBadge from './ids-badge';
import { IdsBlockGrid, IdsBlockGridItem } from './ids-block-grid';
import IdsButton from './ids-button';
import IdsCounts from './ids-counts';
import IdsContainer from './ids-container';
import IdsCard from './ids-card';
import IdsCheckbox from './ids-checkbox';
import IdsColor from './ids-color';
import IdsColorPicker from './ids-color-picker';
import IdsDraggable from './ids-draggable';
import IdsExpandableArea from './ids-expandable-area';
import IdsHeader from './ids-header';
import IdsHyperlink from './ids-hyperlink';
import IdsIcon from './ids-icon';
import IdsInput from './ids-input';
import { IdsLayoutGrid, IdsLayoutGridCell } from './ids-layout-grid';
import IdsLoadingIndicator from './ids-loading-indicator';
import IdsMenuButton from './ids-menu-button';
import IdsModal, { IdsOverlay } from './ids-modal';
import IdsPager, {
  IdsPagerSection,
  IdsPagerInput,
  IdsPagerButton,
  IdsPagerNumberList
} from './ids-pager';
import IdsPopup from './ids-popup';
import IdsProgressBar from './ids-progress-bar';
import IdsRadio from './ids-radio';
import IdsRating from './ids-rating';
import IdsSeparator from './ids-separator';
import IdsSpinbox from './ids-spinbox';
import IdsSwipeAction from './ids-swipe-action';
import IdsSwitch from './ids-switch/ids-switch';
import IdsTabs, { IdsTab, IdsTabsContext, IdsTabContent } from './ids-tabs';
import IdsTag from './ids-tag/ids-tag';
import IdsToggleButton from './ids-toggle-button';
import IdsToolbar, { IdsToolbarSection, IdsToolbarMoreActions } from './ids-toolbar';
import IdsTooltip from './ids-tooltip';
import IdsToast from './ids-toast';
import IdsText from './ids-text';
import IdsTextarea from './ids-textarea';
import { IdsTriggerField, IdsTriggerButton } from './ids-trigger-field';
import IdsUpload from './ids-upload';
import IdsUploadAdvanced from './ids-upload-advanced';
import IdsThemeSwitcher from './ids-theme-switcher';
import IdsWizard, { IdsWizardStep } from './ids-wizard';
