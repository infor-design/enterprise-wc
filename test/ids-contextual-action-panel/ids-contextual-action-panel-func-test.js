/**
 * @jest-environment jsdom
 */
import '../helpers/resize-observer-mock';
import wait from '../helpers/wait';

import IdsContextualActionPanel from '../../src/components/ids-contextual-action-panel';
import { IdsButton } from '../../src/components/ids-button/ids-button';
import { IdsToolbar } from '../../src/components/ids-toolbar';

describe('IdsContextualActionPanel Component', () => {
  let cap;

  beforeEach(async () => {
    const elem = new IdsContextualActionPanel();
    document.body.appendChild(elem);
    cap = document.querySelector('ids-contextual-action-panel');
  });

  afterEach(async () => {
    document.body.innerHTML = '';
    cap = null;
  });

  it('renders with no errors', () => {
    const errors = jest.spyOn(global.console, 'error');
    const elem = new IdsContextualActionPanel();
    document.body.appendChild(elem);
    elem.remove();
    expect(document.querySelectorAll('ids-contextual-action-panel').length).toEqual(1);
    expect(errors).not.toHaveBeenCalled();
  });

  it('renders correctly', () => {
    expect(cap.outerHTML).toMatchSnapshot();
    cap.show();
    expect(cap.outerHTML).toMatchSnapshot();

    cap.hide();
    expect(cap.outerHTML).toMatchSnapshot();
  });

  it('responds to its normal buttons\' clicks', async () => {
    // Insert a header toolbar
    cap.insertAdjacentHTML('afterbegin', `<ids-toolbar slot="toolbar">
      <ids-toolbar-section type="title">
        <ids-text font-size="20" type="h2">Company Information</ids-text>
      </ids-toolbar-section>

      <ids-toolbar-section type="buttonset" align="end">
        <ids-button id="btn-save" icon="save" no-padding>
          <ids-text font-weight="bold">Save</ids-text>
        </ids-button>
        <ids-separator vertical="true"></ids-separator>
        <ids-button id="btn-close" icon="close" no-padding>
          <ids-text font-weight="bold">Close</ids-text>
        </ids-button>
      </ids-toolbar-section>
    </ids-toolbar>`);

    // Setup a button click handler
    cap.popup.animated = false;
    cap.onButtonClick = () => { cap.hide(); };
    const clickEvent = new MouseEvent('click', { bubbles: true });

    // Show the CAP
    await cap.show();
    await wait(310);

    // Click the first CAP button. The above handler should fire.
    cap.querySelector('#btn-save').dispatchEvent(clickEvent);
    await wait(310);

    expect(cap.visible).toBeFalsy();
  });
});
