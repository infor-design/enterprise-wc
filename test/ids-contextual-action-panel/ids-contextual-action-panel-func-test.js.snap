// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`IdsContextualActionPanel Component renders correctly 1`] = `"<ids-contextual-action-panel role=\\"dialog\\"></ids-contextual-action-panel>"`;

exports[`IdsContextualActionPanel Component renders correctly 2`] = `"<ids-contextual-action-panel role=\\"dialog\\" visible=\\"\\" style=\\"z-index: 1020;\\"></ids-contextual-action-panel>"`;

exports[`IdsContextualActionPanel Component renders correctly 3`] = `"<ids-contextual-action-panel role=\\"dialog\\" style=\\"z-index: 1020;\\"></ids-contextual-action-panel>"`;
